### ISSUE TITLE
SimpleRetryStrategy thrashes server if Message + ErrorDescription > Transport Message Limit

### ISSUE THREAD

**Issue by rsnj** [rebus-org/Rebus#626]
We are experiencing an issue where we have a message handler that is throwing an exception with a rather long stack trace. We have the `SimpleRetryStrategy` turned on with the default settings. It correctly retries the handler 5 times and then after it has failed 5 times it correctly tries to move the message with the exception information to the error queue. In this case the stack trace * 5 + the original message is higher than the 65536 byte limit on our Azure Storage Queue transport causing that to fail. So the poisoned message causes an infinite loop on our server bringing the CPU to 100% very quickly and taking it down.

Is it possible to set a limit on the length of the ErrorDescription to prevent this issue?
Is there another handler configuration we are missing that catches these exceptions and correctly handles them?

Rebus 3.1.5

```
16:45:35 An error occurred when attempting to complete the transaction context
Rebus.Exceptions.RebusApplicationException: Could not send message with ID 34f2e89d-8f03-4728-8c6d-01b8c33ca0b6 to 'error' ---> System.ArgumentException: Messages cannot be larger than 65536 bytes.
   at Microsoft.WindowsAzure.Storage.Queue.CloudQueueMessage.GetMessageContentForTransfer(Boolean shouldEncodeMessage, QueueRequestOptions options)
   at Microsoft.WindowsAzure.Storage.Queue.CloudQueue.AddMessageImpl(CloudQueueMessage message, Nullable`1 timeToLive, Nullable`1 initialVisibilityDelay, QueueRequestOptions options)
   at Microsoft.WindowsAzure.Storage.Queue.CloudQueue.BeginAddMessage(CloudQueueMessage message, Nullable`1 timeToLive, Nullable`1 initialVisibilityDelay, QueueRequestOptions options, OperationContext operationContext, AsyncCallback callback, Object state)
   at Microsoft.WindowsAzure.Storage.Core.Util.AsyncExtensions.TaskFromVoidApm[T1,T2,T3,T4,T5](Func`8 beginMethod, Action`1 endMethod, T1 arg1, T2 arg2, T3 arg3, T4 arg4, T5 arg5, CancellationToken cancellationToken)
   at Microsoft.WindowsAzure.Storage.Queue.CloudQueue.AddMessageAsync(CloudQueueMessage message, Nullable`1 timeToLive, Nullable`1 initialVisibilityDelay, QueueRequestOptions options, OperationContext operationContext, CancellationToken cancellationToken)
   at Microsoft.WindowsAzure.Storage.Queue.CloudQueue.AddMessageAsync(CloudQueueMessage message, Nullable`1 timeToLive, Nullable`1 initialVisibilityDelay, QueueRequestOptions options, OperationContext operationContext)
   at Rebus.AzureStorage.Transport.AzureStorageQueuesTransport.<>c__DisplayClass7_0.<<Send>b__0>d.MoveNext()
   --- End of inner exception stack trace ---
   at Rebus.AzureStorage.Transport.AzureStorageQueuesTransport.<>c__DisplayClass7_0.<<Send>b__0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Transport.TransactionContext.<Invoke>d__24.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Transport.TransactionContext.<RaiseCommitted>d__21.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Transport.TransactionContext.<Complete>d__19.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Rebus.Workers.ThreadPoolBased.ThreadPoolWorker.<ProcessMessage>d__18.MoveNext()
```

**Comment by mookid8000** [comment]
Unfortunately this is a pretty hard thing to do anything sensible about.

One could of course truncate the length of the error details to some upper bound, but it would still be possible to hit an unfortunate combination of message body size, stack trace length, and number of retries, which – combined with whichever limit is imposed by the transport – would cause this exception.

If I were you, I would enable compression on message bodies throughout by going

    Configure.With(...)
        .(...)
        .Options(o => o.EnableCompression())
        .Start();

and then try to stay well below the transport's message size limit.

**Comment by mookid8000** [comment]
Now I have though about this for a month, and I haven't come up with anything sensible. Therefore I have added an additional setting in Rebus 4 (currently available as Rebus 4.0.0-b19) where it is possible to limit the length of the error details header like this:
```cs
Configure.With(...)
    .(...)
    .Options(o => o.SimpleRetryStrategy(errorDetailsHeaderMaxLength: 200))
    .Start();
```
which would then truncate all values that exceed 200 characters in length
